### ISSUE TITLE
Template values should be sorted before passing to template system

### ISSUE THREAD

**Issue by RhetTbull** [RhetTbull/osxphotos#1317]
> Could it be that osxphotos while building and comparing metadata gets the albums sorted differently on each run into the description field?

Ahah! That's it! Because of the way the data is loaded from the database the order of multi-value items like keywords and albums is not guaranteed to be consistent as you've discovered.  One solution is to sort them before passing into the template system but I need to think about this to ensure there's no 2nd order effects of doing so. 

The osxphotos template language does provide a `|sort` filter which sorts a list of values so that could be helpful. I'll think about this some more because it was not intuitive that what you're doing should cause the order to change.

Until I get a fix in place, you can sort the output yourself to ensure albums are always in sorted order (this could change a lot of your photos though):

`Album:{album|sort,}{newline}Description:{descr,}`

I'm not sure this template does what you want though. A couple issues: 

- If there's no album it still prints `Album:`
- If there's no description, it still prints `Descr:`
- Most importantly, it prints the description as many times as there are albums because the entire template gets evaluated for the combination of all possible values. Thus, a photo in two albums will get a description like this:

```
  Album:Album 1
  Description:The description
  Album:Album 2
  Description:The description
```

Here's a different template that prints the `Album:` part first (and sorted) then the description. This is more complex but might be a better fit for what you're trying to do:

`{var:NL,{newline}}{var:AlbumDescr,{album?Album:{album|sort,},}}{%AlbumDescr?{%AlbumDescr|join(%NL)},}{descr?{newline}Descr:{descr},}`

This prints out:

```
  Album:Album 1
  Album:Album 2
  Description:The description
```

`{var:NL,{newline}}`: Use `var` variable feature to create a variable called `NL` which holds a newline (needed because templates cannot be used in filter fields but variables can)

`{var:AlbumDescr,{album?Album:{album|sort,},}}`: and a variable called `AlbumDescr` which is the `Album:Album Name` part except that it sorts the albums and only does this if there are albums. If you wanted `Album:` to show up even if no album, you can set the default value to `Album:`

`{%AlbumDescr?{%AlbumDescr|join(%NL)},}`: Test to see if the `%AlbumDescr` variable has a value, if so, use the `|join()` filter to join the values with a `{newline}` but since filters cannot accept templates as arguments, use the `%NL` variable we defined earlier.  The output of this is a single string in form:

```
  Album:Album 1
  Album:Album 2
```

`{descr?{newline}Descr:{descr},}`: Test to see if `{descr}` has a value if so, add a new line then add the `Descr:description` part. 

This adds a new line before description even when there's no album -- out of time at the moment but can go back and tweak later to fix this.

_Originally posted by @RhetTbull in https://github.com/RhetTbull/osxphotos/issues/1311#issuecomment-1847449560_

**Comment by RhetTbull** [comment]
Looking at the code I see some values like keywords and labels are sorted but albums are not (though `{folder_album}` is).  All multi-value template values should be sorted.

https://github.com/RhetTbull/osxphotos/blob/d81c58f58d7bf69b25bc7a15ca96f5fe86a3f69b/osxphotos/phototemplate.py#L1213-L1233

**Comment by RhetTbull** [comment]
Working on a fix but this may cause sidecars to be regenerated for those using `--sidecar`.

**Comment by RhetTbull** [comment]
Fixed in v0.66.0
